cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -474,7 +572,36 @@ func (r *ControllerDescriptor) RequiresSpecialHandling() bool {
 	return r.requiresSpecialHandling
 }
 
-// KnownControllers returns all known controllers's name
+// BuildController creates a controller based on the given descriptor.
+// The associated controller's constructor is called at the end, so the same contract applies for the return values here.
+func (r *ControllerDescriptor) BuildController(ctx context.Context, controllerCtx ControllerContext) (Controller, error) {
+	logger := klog.FromContext(ctx)
+	controllerName := r.Name()
+
+	for _, featureGate := range r.GetRequiredFeatureGates() {
+		if !utilfeature.DefaultFeatureGate.Enabled(featureGate) {
+			logger.Info("Controller is disabled by a feature gate",
+				"controller", controllerName,
+				"requiredFeatureGates", r.GetRequiredFeatureGates())
+			return nil, nil
+		}
+	}
+
+	if r.IsCloudProviderController() {
+		logger.Info("Skipping a cloud provider controller", "controller", controllerName)
+		return nil, nil
+	}
```

Comment:
It's too bad we're propagating the cloud provider extraction legacy into new code.  Anything we can do to avoid that or further minimize it? (doesn't have to be blocking for this PR.. feel free to open an issue if there are ideas but we don't have the energy to deal with them now)

Reply:
I'll file an issue for the cloud provider extraction legacy code.